text/0000-async-work-service.md

```diff
@@ -0,0 +1,322 @@
+
+
+
+### Use Case 3: Unbounded Asynchronous Work
+
+Jeff is a CppMicroServices maintainer responsible for CppMicroServices and it's compendium services. All of the asynchronous tasks done in CppMicroServices core framework, Declarative Services and Config Admin are unbounded and need to be waited upon. Most of the asynchronous tasks involve calling back into user code via callbacks and as such cannot be guaranteed to finish in a bounded amount of time. Currently, `std::async` and `boost::asio::post` are being used to execute unbounded async tasks, both of which use futures for the calling thread to block on. Any solution which replaces `std::async` or `boost::asio::post` needs to provide a way to wait on the result and a way to receive a failure from the async task (**PP3**). 
+
+
+
+### Requirements
+
+| ID   | Statement                                                    | Source / Pain Point | Priority  |
+| ---- | ------------------------------------------------------------ | ------------------- | --------- |
+| 1    | The service interface should be re-usable by any other CppMicroServices service or application. | PP1                 | Must Have |
+| 2    | The service interface should be decoupled from CppMicroServices and all compendium services. | PP2                 | Must Have |
+| 3    | CppMicroServices service based design                        | OSGi compliance     | Must Have |
+| 4    | Allow users to block on the asynchronous task until it has finished. | PP3                 | Must Have |
+| 5    | Allow users to receive a failure from the asynchronous task. | PP3                 | Must Have |
+
+
+
+## Detailed design
+
+> This is the bulk of the RFC.
+
+> Explain the design in enough detail for somebody
+familiar with the framework to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+
+
+### API
+
+```c++
+namespace cppmicroservices { 
+  namespace async {
+    /**
+     *
+     */
+    class UnboundedAsyncTask {
+    public:
+      virtual ~UnboundedAsyncTask() noexcept = default;
+      
+      /**
+       * Run a std::function<void()> object asynchronously on another thread.
+       *
+       * @param task A std::function<void()> object representing a Callable target.
+       * @return A std::future<void> object to wait on the result of task.
+       * @note There is no guarantee made by this interface that the 
+       * std::future<void> returned will automatically block when it goes out of scope. 
+       * It may or may not depending on the implementation of this method.
+       */
```

Comment:
be more precise on the requirements of the API. the way it's worded `std::async` has to be used to implement this API. 
- what happens if work is dependent on one another? would a thread pool of size 1 be sufficient?
- would inline execution be acceptable?